lib/src/main/java/edu/stanford/futuredata/macrobase/analysis/classify/MultiMADClassifier.java

```diff
@@ -0,0 +1,291 @@
+package edu.stanford.futuredata.macrobase.analysis.classify;
+
+import edu.stanford.futuredata.macrobase.datamodel.DataFrame;
+import edu.stanford.futuredata.macrobase.operator.Transformer;
+
+import org.apache.commons.math3.special.Erf;
+import org.apache.commons.math3.stat.descriptive.rank.Percentile;
+
+import java.lang.Math;
+
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Random;
+
+/**
+ * This class processes a batch of rows. For each input metric column, it classifies the rows as
+ * outliers or inliers based on the median absolute deviation (MAD) from the median, and writes
+ * the classification results to a new output column. Each input column is considered as an
+ * independent unit. For example, if there are five input metric columns, each row is classified
+ * five times, once for each column, and five new ouput columns will be created.
+ */
+public class MultiMADClassifier implements Transformer {
+    private double percentile = 0.5;
+    private double zscore = 2.576;
+    private boolean includeHigh = true;
+    private boolean includeLow = true;
+    private double samplingRate = 1;
+    private List<String> columnNames;
+    private String outputColumnSuffix = "_OUTLIER";
+    private boolean useParallel = false;
+    private boolean useReservoirSampling = false;
+    private final double trimmedMeanFallback = 0.05;
+    // https://en.wikipedia.org/wiki/Median_absolute_deviation#Relation_to_standard_deviation
+    private final double MAD_TO_ZSCORE_COEFFICIENT = 1.4826;
+
+    // Calculated values
+    private DataFrame output;
+
+    public MultiMADClassifier(String... columnNames) {
+        this.columnNames = new ArrayList<String>(Arrays.asList(columnNames));
+    }
+
+    @Override
+    public void process(DataFrame input) {
+        output = input.copy();
+
+        int sampleSize = (int)(input.getNumRows() * samplingRate);
+        int[] sampleIndices = new int[sampleSize];
```

Comment:
Is this necessary (allocating a new array) when sampleRate == 1.0? might as well avoid extra work. In general I think it would be nicer if you could factor out all of the "sampling" logic more cleanly rather than mixing it inline in multiple places, see what you can do.